### PR TITLE
doc: add plain+v0 bundle spec

### LINKED ISSUE #75: Document the plain manifest bundle format
Goal: document the plain manifest bundle format and the required Dockerfile format.

Motivation: #71 introduces user documentation for interacting with the plain bundle provisioner, and we're missing the ability to link to documentation that acts as a source-of-truth for this bundle format.

### PR DESCRIPTION
**Pull request by exdx**
Closes #75 .

Additional metadata (as discussed in #23) is being treated as optional in the current plain+v0 spec, since there are no provisioner specific requirements around metadata at this time. This spec omits metadata until we agree on a specified metadata format and the benefits of requiring such a format.

### CODE REVIEW

File: docs/plain-bundle-spec.md
Comment (akihikokuroda) [inline docs/plain-bundle-spec.md:11]: Where is the current schema version is defined (or specified)?

File: docs/plain-bundle-spec.md
Comment (akihikokuroda) [inline docs/plain-bundle-spec.md:37]: Can the manifests directory have sub-directories that also have yaml files?

File: docs/plain-bundle-spec.md
Comment (exdx) [reply docs/plain-bundle-spec.md:11]: This is a good question. I'm not sure that we have a formal definition anywhere of how we version the schema. The proposed schema version `v0` is meant to indicate that the schema version is experimental and still not fully set in stone. We can clarify this more in the doc.

File: docs/plain-bundle-spec.md
Comment (exdx) [inline docs/plain-bundle-spec.md:99]: nit: this needs to be updated once #160 lands.

File: internal/provisioner/plain/README.md
Comment (exdx) [inline internal/provisioner/plain/README.md:10]: The changes here and below in this file are all as a result of markdown linting.

Comment (exdx) [inline internal/provisioner/plain/README.md:10]: The changes here and below in this file are all just linting modifications.

File: docs/plain-bundle-spec.md
Comment (timflannagan) [inline docs/plain-bundle-spec.md:87]: nit: bash

File: docs/plain-bundle-spec.md
Comment (timflannagan) [inline docs/plain-bundle-spec.md:81]: nit: bash

File: docs/plain-bundle-spec.md
Comment (timflannagan) [inline docs/plain-bundle-spec.md:61]: Duplication compared to the plain provisioner README?

File: docs/plain-bundle-spec.md
Comment (timflannagan) [inline docs/plain-bundle-spec.md:58]: I hate to be a broken record: can we remove these labels as they're essentially a no-op right now (and there's no concrete evidence, at least for now, that these labels will be needed)?

Comment (timflannagan) [reply docs/plain-bundle-spec.md:58]: If we're interesting in attaching metadata right now to an individual Bundle, maybe use labels/annotations on the actual Bundle resource?

**Review by timflannagan — COMMENTED** [review]
I think this is getting closer once we rebase this against the latest master changes now that the Bundle API contains a source type.

In general, I think I had the following outline in mind:

- What is this bundle format
- What are the support source types for this bundle format?
- How can I use this format and/or quickstart
- (optional) How can I validate this bundle format is valid?

WDYT?

File: docs/plain-bundle-spec.md
Comment (exdx) [reply docs/plain-bundle-spec.md:58]: Ok, no problem, I'll take them out -- I think the reason I'm so interested in the labels is because they are used heavily in the `registry+v1` format and I think they may eventually be useful. But as you stated, there's no evidence that they are useful for the plain+v0 format right now and we can drop them for now.

Comment (timflannagan) [reply docs/plain-bundle-spec.md:58]: Yep - this can always be an additive change if we ever find the need for exposing container-level labels like we do with the registry+v1 format.

**Comment by exdx** [comment]
> 

That sounds good. Yes, the validation is something that we are currently missing and may want to implement soon. Even though the format is pretty straightforward, having a tool that can validate up-front may be useful.

File: docs/plain-bundle-spec.md
Comment (joelanford) [inline docs/plain-bundle-spec.md:15]: I've been thinking about our naming. I wonder if we should reuse the `metav1.Object` concept of using apiVersion and kind (e.g. `apiVersion: core.rukpak.io/v0, kind: Plain`).

It gives us everything we need:
- explicit namespacing with a domain name-based group name
- major versioning
- different kinds of bundles
- extensibility (anyone can create a bundle format and a provisioner in their own group)

It also gives us a more structured type to store all of this metadata, so that we don't have to argue over "is it plain.v0, plain+v0, or plain-v0?"

File: docs/plain-bundle-spec.md
Comment (joelanford) [inline docs/plain-bundle-spec.md:68]: Should we define some terms somewhere?
- `bundle` is a collection of files that define content to be deployed to a cluster.
- `bundle image` is a container image that contains a bundle (any reason the bundle has to be at the image root? we could augment the `image` source type to allow specifying a subdirectory in the image, defaulting to `/manifests`, just like we're doing for bundle repos)
- `bundle git repo` is a git repo that contains a bundle (where the bundle can be anywhere within the repo)

File: docs/plain-bundle-spec.md
Comment (joelanford) [inline docs/plain-bundle-spec.md:69]: I'd suggest that we start with a flat directory for now. We can always allow subdirectories in the future without making a breaking change, but we can't go the other way around. Supporting subdirectories might complicate the ordering conversation we were having earlier.

File: docs/plain-bundle-spec.md
Comment (joelanford) [inline docs/plain-bundle-spec.md:73]: IMO, we should not allow metadata to be provided in different ways. That introduces unnecessary ambiguity and complexity:
- do I have to include both?
- if I include both, which takes precedence?
- is precedence always the same no matter the context (e.g. building a catalog vs unpacking the bundle on cluster)?

My vote: Only metadata on the filesystem. That ensures that all filesystem-based sources can provide metadata in a consistent way. The downside is that for images, you have to pull the image to get metadata, rather than just requesting labels. But IMO the tradeoff is easily worth it, at least given my personal experience and confusion with registry+v1 metadata.

For bundle metadata, I think we should define a single file (`bundle.yaml`) that must exist in the root of _ALL_ bundle directories for all bundles types (that would mean we'd need special handling for registry+v1 bundles). And the only required fields would be `apiVersion` and `kind`.  That way any provisioner can look at any bundle and immediately parse this one file to definitely identify what it is looking at.

File: docs/plain-bundle-spec.md
Comment (joelanford) [inline docs/plain-bundle-spec.md:75]: We should try to be as specific as we can here. For example:
- Are JSON manifests allowed?
- Are multiple objects allowed in a single file?
- Are specific file extensions expected?

This doc should include instructions (maybe TODO for now) for how someone can build and/or validate a plain+v0 bundle.

File: docs/plain-bundle-spec.md
Comment (joelanford) [inline docs/plain-bundle-spec.md:110]: Use heredoc syntax so folks following along can copy paste into their terminal?

File: docs/plain-bundle-spec.md
Comment (joelanford) [inline docs/plain-bundle-spec.md:138]: Same here re: heredoc (and perhaps add the `kubectl apply` command?)

File: docs/plain-bundle-spec.md
Comment (timflannagan) [reply docs/plain-bundle-spec.md:68]: Yeah I think adding a common terminology section might help make this easier to reason about over time?

File: docs/plain-bundle-spec.md
Comment (timflannagan) [reply docs/plain-bundle-spec.md:69]: +1

File: docs/plain-bundle-spec.md
Comment (timflannagan) [reply docs/plain-bundle-spec.md:73]: Do we even needs to document that we have support (or rather, will have support down the line) for exposing bundle metadata? Can this be added over time once we have a more concrete use case for bundle-level metadata?

Comment (joelanford) [reply docs/plain-bundle-spec.md:73]: Yeah, if we completely remove this piece, that also satisfies me for now.

File: docs/plain-bundle-spec.md
Comment (exdx) [reply docs/plain-bundle-spec.md:15]: I do like this idea, I think it's intuitive. Using `Kind` to specify different types of Bundles is nice UX too.

File: docs/plain-bundle-spec.md
Comment (timflannagan) [inline docs/plain-bundle-spec.md:8]: nit: broken link.

File: docs/plain-bundle-spec.md
Comment (timflannagan) [inline docs/plain-bundle-spec.md:13]: nit: broken link.

File: docs/plain-bundle-spec.md
Comment (timflannagan) [inline docs/plain-bundle-spec.md:17]: nit: broken link.

File: docs/plain-bundle-spec.md
Comment (joelanford) [reply docs/plain-bundle-spec.md]: My suggestion for combo for now would be:

```
combo
└── bundle
    └── manifests
        └── manifest.yaml
```

in case we put anything else in the bundle spec that would be a sibling of `manifests`. If we collapse it down and remove the `bundle` directory from the tree, then there's potential for bundle metadata stuff to be intermingled with other operator source code.

That's kind of why I like the idea of `bundle.yaml` with that name; it _could_ live at that root level and not be some non-descript file sitting there.

File: docs/plain-bundle-spec.md
Comment (joelanford) [inline docs/plain-bundle-spec.md:76]:
```suggestion
* Including any content in the root `manifests` directory of a plain bundle that is not static manifests will result in a failure when creating content
```

File: docs/plain-bundle-spec.md
Comment (joelanford) [reply docs/plain-bundle-spec.md:133]: Nvmd: I think I see what's happening. We're creating a plain manifest bundle for the provisioner, then we're installing the provisioner separately with `make run`, then we're creating a `Bundle` so that we can see the contents of the provisioner's manifests.

It's pretty confusing IMO to use the plain provisioner's manifests for the bundle image.

Comment (exdx) [reply docs/plain-bundle-spec.md:133]: That's true, I partially chose it because it meant the user didn't have to go and pull down another repository or set of files in order to build a bundle.

We could potentially use combo but it's `manifests` directory is not flat and so it doesn't adhere to the spec. 

I made https://github.com/exdx/combo-bundle for testing purposes, we could point to that instead, WDYT?

Comment (timflannagan) [reply docs/plain-bundle-spec.md:133]: Can we simply update combo and deflate everything to a single directory?

**Review by timflannagan — APPROVED** [review]
I think this is ready to go as an initial stab, and we can continue to iterate on the content/structure over time. Note: I _think_ there's still an issue with the plain-bundle-spec.md containing broken links still when I use the rendered markdown view, and individual click on those links, but I'll let you decide whether this is needed, and if so, can be done as a quick follow-up.

Let's get another reviewer though to approve. Maybe @joelanford if he has a chance today?

**Review by joelanford — APPROVED** [review]

**Pull request merged**
